net: lwm2m: Bootstrap procedure improvements

rlubos opened the pull request:
This PR introduces several improvements and fixes for LwM2M bootstrap procedure. Detailed description of the changes can be found in the commit message, to summarize:

Improvements:
 * Introduced new state machine state, `ENGINE_IDLE`, which now is the default after boot and after LwM2M engine is stopped.
 * Allow to specify whether bootstrap procedure shall run in the current session, instead of relying on static configuration only.

Fixes:
 * Introduce new member in the resource instance structure, indicating actual data size stored (so far only the buffer size was kept, making it impossible to retrieve the actual opaque data size written by the bootstrap server).
 * Fix PSK ID provisioning, by using the actual data size instead of result of `strlen()` (the PSK ID written by the server is of opaque data type and is not NULL terminated).
 * Report Bootstrap Complete event only after all bootstrap transactions are done.

Based on https://github.com/zephyrproject-rtos/zephyr/pull/28592, marking DNM until it's merged.

rlubos committed:
net: lwm2m: Introduce ENGINE_IDLE state

So far, the LwM2M state machine started in the `ENGINE_INIT` state,
which made it exectue the registration/bootstrap registration even when
`lwm2m_rd_client_start()` was not called. With a new `ENGINE_IDLE`
state, the state machine can wait for the application to actually start
the client before proceeding. It also makes sense to stay in the
ENGINE_IDLE state after successfull deregistration, until the
application restarts the client.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Make bootstrap optional

Currently, after `CONFIG_LWM2M_RD_CLIENT_SUPPORT_BOOTSTRAP` is enabled,
the LwM2M engine will initiate bootstrap procedure on each run. This
approach limits the flexibility of the application, as it's not always
necessary to go over the bootstrap procedure (for instance, the
application may decide to store the security object obtained during the
bootstrap in flash, and restore it on boot).

Fix this by introducing an additional `flags` parameter to the
`lwm2m_rd_client_start()` function, which provides information whether
to run bootstrap in the current session or not.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Store the actual resource size in the resource instance

So far, the resource instance structure kept only the information about
the buffer length provided to the resource (in the `data_len` field).
While this approach might be enough for integer resources, where the
actual data size is fixed, it did not work for opaque resources. It is
impossible to determine the actual opaque resource length after it's
been written into.

Fix this, by replacing the current `data_len` field of the
`lwm2m_engine_res_inst` with `max_data_len`, indicating the buffer
size, and making the `data_len` field to hold the actual data size of
the resource.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Report boostrap complete after final response is sent

So far, `LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_TRANSFER_COMPLETE` event was
reported before the final ACK for the Bootstrap Finish was sent from the
client side. This could cause delays in the ACK sending, in case the
application wanted for instance to store the received data in flash.

Fix this, by reporting the
`LWM2M_RD_CLIENT_EVENT_BOOTSTRAP_TRANSFER_COMPLETE` event on the next
state tansition (before the actual registration starts).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Use the actual data size when provisioning PSK ID

"Public Key or Identity" resource is of opaque data type, therefore it's
not correct to assume it will be a NULL terminated string (the existing
servers, for instance Leshan, does not include NULL terminator). Use the
actual size associated with the resource instead.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

jukkar reviewed (approved):
LGTM

carlescufi commented on include/net/lwm2m.h:
```diff
@@ -871,12 +871,13 @@ typedef void (*lwm2m_ctx_event_cb_t)(struct lwm2m_ctx *ctx,
  */
 void lwm2m_rd_client_start(struct lwm2m_ctx *client_ctx, const char *ep_name,
-			   lwm2m_ctx_event_cb_t event_cb);
+			   bool bootstrap, lwm2m_ctx_event_cb_t event_cb);
```
Let's use an `flags` integer instead so we are future-proof.

rlubos replied:
I've added `flags` parameter as suggested and a release notes entry.

jukkar commented on include/net/lwm2m.h:
```diff
@@ -871,12 +880,13 @@ typedef void (*lwm2m_ctx_event_cb_t)(struct lwm2m_ctx *ctx,
  *
  * @return 0 for success or negative in case of error.
  */
 void lwm2m_rd_client_start(struct lwm2m_ctx *client_ctx, const char *ep_name,
```
Just noticed this, the function is void but comment says it returns 0 for success and <0 for error. Which one is correct?

rlubos replied:
It's void, I'll remove the invalid docs.

rlubos committed:
net: lwm2m: Remove invalid doc entries

RD client APIs don't return anything, yet return values were documented.
Remove invalid docs.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
doc: reference: Add separate entries for CoAP, LwM2M and MQTT APIs

Application protocol APIs are unrelated to each other and change
separately, so it makes sense to have separate entries for each of them.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar I've removed the invalid API doc entries.

@carlescufi I've added separate entries for application protocol APIs (CoAP, MQTT, LwM2M) to the overview page, as requested. Not sure if there's any pattern in adding these entries (they're not sorted alphabetically), for now I've added the new entries just below "Networking".
